pkg/cache/v3/linear.go

```diff
@@ -239,6 +239,45 @@ func (cache *LinearCache) CreateWatch(request *Request) (chan Response, func())
 	}
 }
 
+// SetResources replaces current resources with a new set of resources
+//
+// If we have individual watches for two resources (there are 2 DiscoveryRequest with resourceNames=res1 and resourceNames=res2)
+// Then if you set new resources, res1 changed but res2 is not, we assume that all resources are changed so both watches will be triggered.
+// That's why this function is well suited for a use case when resourceNames of DiscoveryRequest is empty.
+// This way you don't have to call UpdateResource for every resource which will trigger watch of DiscoveryRequest(resourceNames="") twice,
+// but it will only happen once regardless of how many resources are changed.
+func (cache *LinearCache) SetResources(resources map[string]types.Resource) {
+	cache.mu.Lock()
+	defer cache.mu.Unlock()
+
+	cache.version += 1
+
+	modified := map[string]struct{}{}
+	for name := range cache.resources {
+		if _, found := resources[name]; !found {
+			delete(cache.versionVector, name)
+			modified[name] = struct{}{}
+		}
+	}
+
+	cache.resources = resources
+	for name := range resources {
+		// We assume all resources passed to SetResources are changed.
+		// Otherwise we would have to do proto.Equal on resources which is pretty expensive operation
+		cache.versionVector[name] = cache.version
+		modified[name] = struct{}{}
+	}
+
+	cache.notifyAll(modified)
+}
+
+// GetResources returns current resources stored in the cache
+func (cache *LinearCache) GetResources() map[string]types.Resource {
+	cache.mu.Lock()
+	defer cache.mu.Unlock()
+	return cache.resources
+}
+
```

Comment:
Just a nit pick but can we move both these functions above the `CreateWatch()`? I'd like to keep them bundled with Update/DeleteResource().